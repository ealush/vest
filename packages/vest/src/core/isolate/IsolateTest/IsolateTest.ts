import { CB, Maybe } from 'vest-utils';
import { TIsolate, Isolate, IsolateKey } from 'vestjs-runtime';

import { TestStatus } from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { AsyncTest, TestFn } from 'TestTypes';
import { VestIsolateType } from 'VestIsolateType';

export type TIsolateTest<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
> = TIsolate<CommonTestFields<F, G> & IsolateTestPayload>;

export function IsolateTest<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
>(
  callback: CB,
  input: CommonTestFields<F, G>,
  key?: IsolateKey
): TIsolateTest<F, G> {
  const payload: IsolateTestPayload = {
    ...IsolateTestBase(),
    fieldName: input.fieldName,
    testFn: input.testFn,
  };

  if (input.groupName) {
    payload.groupName = input.groupName;
  }

  if (input.message) {
    payload.message = input.message;
  }
  const isolate = Isolate.create<IsolateTestPayload>(
    VestIsolateType.Test,
    callback,
    payload,
    key ?? null
  );

  return isolate as TIsolateTest<F, G>;
}

export function IsolateTestBase() {
  return {
    severity: TestSeverity.Error,
    status: TestStatus.UNTESTED,
    abortController: new AbortController(),
  };
}

export type IsolateTestPayload<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
> = CommonTestFields<F, G> & {
  severity: TestSeverity;
  status: TestStatus;
  asyncTest?: AsyncTest;
  abortController: AbortController;
};

type CommonTestFields<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
> = {
  message?: Maybe<string>;
  groupName?: G;
  fieldName: F;
  testFn: TestFn;
};
