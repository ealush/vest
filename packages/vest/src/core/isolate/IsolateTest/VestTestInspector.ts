import { isPromise } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import { TestStatus } from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';

export class VestTestInspector {
  static warns(test: TIsolateTest): boolean {
    return test.severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return (
      VestTestInspector.isFailing(test) || VestTestInspector.isWarning(test)
    );
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTestInspector.isSkipped(test) ||
      VestTestInspector.isOmitted(test) ||
      VestTestInspector.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return (
      VestTestInspector.hasFailures(test) || VestTestInspector.isPassing(test)
    );
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTestInspector.isSkipped(test) ||
      VestTestInspector.isUntested(test) ||
      VestTestInspector.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(test.asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return test.status === status;
  }
}
