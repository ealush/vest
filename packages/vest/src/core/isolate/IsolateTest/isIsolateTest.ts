import { Maybe, invariant } from 'vest-utils';
import { TIsolate, IsolateSelectors } from 'vestjs-runtime';

import type { TIsolateTest } from 'IsolateTest';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { VestIsolateType } from 'VestIsolateType';

export function isIsolateTest(
  isolate?: Maybe<TIsolate>
): isolate is TIsolateTest {
  return IsolateSelectors.isIsolateType<TIsolateTest>(
    isolate,
    VestIsolateType.Test
  );
}

export function isIsolateTestX(
  isolate: TIsolate
): asserts isolate is TIsolateTest {
  invariant(isIsolateTest(isolate));
}

export function castIsolateTest<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
>(isolate: TIsolate): TIsolateTest<F, G> {
  isIsolateTestX(isolate);

  return isolate as TIsolateTest<F, G>;
}
