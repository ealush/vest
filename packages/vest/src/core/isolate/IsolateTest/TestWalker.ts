import { Walker, Isolate, VestRuntime } from 'vest-runtime';

import { IsolateTest } from 'IsolateTest';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import matchingFieldName from 'matchingFieldName';

type MaybeRoot = Isolate | null;

export class TestWalker {
  static defaultRoot() {
    return VestRuntime.useAvailableRoot();
  }

  static hasNoTests(root: MaybeRoot = TestWalker.defaultRoot()): boolean {
    if (!root) return true;
    return !Walker.has(root, IsolateTest.is);
  }

  static someIncompleteTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.some(
      root,
      isolate => {
        IsolateTest.isX(isolate);

        return isolate.isPending() && predicate(isolate);
      },
      IsolateTest.is
    );
  }

  static someTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.some(
      root,
      isolate => {
        IsolateTest.isX(isolate);

        return predicate(isolate);
      },
      IsolateTest.is
    );
  }

  static everyTest(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.every(
      root,
      isolate => {
        IsolateTest.isX(isolate);

        return predicate(isolate);
      },
      IsolateTest.is
    );
  }

  static walkTests<F extends TFieldName, G extends TGroupName>(
    callback: (test: IsolateTest<F, G>, breakout: () => void) => void,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    if (!root) return;
    Walker.walk(
      root,
      (isolate, breakout) => {
        callback(IsolateTest.cast<F, G>(isolate), breakout);
      },
      IsolateTest.is
    );
  }

  static hasRemainingTests(fieldName?: TFieldName): boolean {
    return TestWalker.someIncompleteTests(testObject => {
      if (fieldName) {
        return matchingFieldName(testObject, fieldName);
      }
      return true;
    });
  }

  static pluckTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    if (!root) return;
    Walker.pluck(
      root,
      isolate => {
        IsolateTest.isX(isolate);

        return predicate(isolate);
      },
      IsolateTest.is
    );
  }

  static resetField(fieldName: TFieldName): void {
    TestWalker.walkTests(testObject => {
      if (matchingFieldName(testObject, fieldName)) {
        testObject.reset();
      }
    }, TestWalker.defaultRoot());
  }

  static removeTestByFieldName(
    fieldName: TFieldName,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    TestWalker.pluckTests(testObject => {
      return matchingFieldName(testObject, fieldName);
    }, root);
  }
}
