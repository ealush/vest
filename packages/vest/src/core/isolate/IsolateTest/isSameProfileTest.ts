import { TIsolateTest } from 'IsolateTest';
import matchingFieldName from 'matchingFieldName';

export function isSameProfileTest(
  testObject1: TIsolateTest,
  testObject2: TIsolateTest
): boolean {
  return (
    matchingFieldName(testObject1, testObject2.fieldName) &&
    testObject1.groupName === testObject2.groupName &&
    testObject1.key === testObject2.key
  );
}
