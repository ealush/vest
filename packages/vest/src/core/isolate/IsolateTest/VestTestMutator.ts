import { IsolateInspector, IsolateMutator } from 'vestjs-runtime';

import { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { VestTestInspector } from 'VestTestInspector';

const TestStateMachine = createTestStateMachine();

export class VestTestMutator {
  static setPending(test: TIsolateTest) {
    VestTestMutator.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTestMutator.setStatus(
      test,
      VestTestInspector.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTestMutator.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    IsolateMutator.setData(test, 'severity', TestSeverity.Warning);
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTestMutator.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTestMutator.setStatus(test, TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTestMutator.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTestMutator.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    IsolateMutator.setData(
      test,
      'status',
      TestStateMachine.staticTransition(
        IsolateInspector.getData(test).status,
        status,
        payload
      )
    );
  }
}
