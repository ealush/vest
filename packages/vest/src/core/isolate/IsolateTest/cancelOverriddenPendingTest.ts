import { TIsolateTest } from 'IsolateTest';
import { VestTest, VestTest } from 'VestTest';
import { isSameProfileTest } from 'isSameProfileTest';

export default function cancelOverriddenPendingTest(
  prevRunTestObject: TIsolateTest,
  currentRunTestObject: TIsolateTest
): void {
  if (
    currentRunTestObject !== prevRunTestObject &&
    isSameProfileTest(prevRunTestObject, currentRunTestObject) &&
    VestTest.isPending(prevRunTestObject)
  ) {
    VestTest.cancel(prevRunTestObject);
  }
}
