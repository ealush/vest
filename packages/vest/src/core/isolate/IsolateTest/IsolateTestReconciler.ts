import { ErrorStrings } from 'ErrorStrings';
import { Reconciler } from 'vest-runtime';
import type { Isolate } from 'vest-runtime';
import { deferThrow, isNullish, text } from 'vest-utils';

import { IsolateTest } from 'IsolateTest';
import cancelOverriddenPendingTest from 'cancelOverriddenPendingTest';
import { isSameProfileTest } from 'isSameProfileTest';
import { useVerifyTestRun } from 'verifyTestRun';

// @vx-allow use-use
export function IsolateTestReconciler(
  currentNode: Isolate,
  historyNode: Isolate | null
): Isolate {
  // Start by verifying params
  if (!IsolateTest.is(currentNode)) {
    // This is unreachable, since this function should only be called with IsolateTest nodes
    return currentNode;
  }

  if (isNullish(historyNode)) {
    return handleNoHistoryNode(currentNode);
  }

  if (!IsolateTest.is(historyNode)) {
    return currentNode;
  }

  const reconcilerOutput = usePickNode(historyNode, currentNode);

  cancelOverriddenPendingTestOnTestReRun(
    reconcilerOutput,
    currentNode,
    historyNode
  );

  return reconcilerOutput;
}

// eslint-disable-next-line max-statements

function nodeReorderDetected(
  newNode: IsolateTest,
  prevNode?: Isolate
): boolean {
  return !!IsolateTest.is(prevNode) && !isSameProfileTest(prevNode, newNode);
}

function handleCollision(
  newNode: IsolateTest,
  prevNode?: Isolate
): IsolateTest {
  if (newNode.usesKey()) {
    return IsolateTest.cast(Reconciler.handleIsolateNodeWithKey(newNode));
  }

  if (nodeReorderDetected(newNode, prevNode)) {
    return onNodeReorder(newNode, prevNode);
  }

  if (!IsolateTest.is(prevNode)) {
    // I believe we cannot actually reach this point.
    // Because it should already be handled by nodeReorderDetected.
    return newNode;
  }

  // FIXME: May-13-2023
  // This may not be the most ideal solution.
  // In short: if the node was omitted in the previous run,
  // we want to re-evaluate it. The reason is that we may incorrectly
  // identify it is "optional" because it was omitted in the previous run.
  // There may be a better way to handle this. Need to revisit this.
  if (prevNode.isOmitted()) {
    return newNode;
  }

  return prevNode;
}

function onNodeReorder(newNode: IsolateTest, prevNode?: Isolate): IsolateTest {
  throwTestOrderError(newNode, prevNode);
  Reconciler.removeAllNextNodesInIsolate();
  return newNode;
}

function usePickNode(
  historyNode: IsolateTest,
  currentNode: IsolateTest
): Isolate {
  const collisionResult = handleCollision(currentNode, historyNode);

  return useVerifyTestRun(currentNode, collisionResult);
}

function handleNoHistoryNode(testNode: IsolateTest): IsolateTest {
  if (testNode.usesKey()) {
    return IsolateTest.cast(Reconciler.handleIsolateNodeWithKey(testNode));
  }

  return testNode;
}

function cancelOverriddenPendingTestOnTestReRun(
  nextNode: Isolate,
  currentNode: Isolate,
  prevTestObject: IsolateTest
) {
  if (nextNode === currentNode && IsolateTest.is(currentNode)) {
    cancelOverriddenPendingTest(prevTestObject, currentNode);
  }
}

function throwTestOrderError(
  newNode: IsolateTest,
  prevNode: Isolate | undefined
): void {
  if (newNode.shouldAllowReorder()) {
    return;
  }

  deferThrow(
    text(ErrorStrings.TESTS_CALLED_IN_DIFFERENT_ORDER, {
      fieldName: newNode.fieldName,
      prevName: IsolateTest.is(prevNode) ? prevNode.fieldName : undefined,
    })
  );
}
