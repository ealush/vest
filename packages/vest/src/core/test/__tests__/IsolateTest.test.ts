import { noop } from 'vest-utils';
import wait from 'wait';

import { TestPromise } from '../../../../testUtils/testPromise';

import { IsolateTest, TIsolateTest } from 'IsolateTest';
import { VestTestInspector } from 'VestTestInspector';
import { VestTestMutator } from 'VestTestMutator';
import * as vest from 'vest';

const fieldName = 'unicycle';
const message = 'I am Root.';

describe('IsolateTest', () => {
  let testObject: TIsolateTest;

  beforeEach(() => {
    testObject = IsolateTest(noop, {
      fieldName,
      testFn: jest.fn(),
      message,
    });
  });

  test('TestObject constructor', () => {
    expect(testObject).toMatchSnapshot();
  });

  it('Should have a unique id', () => {
    Array.from({ length: 100 }, () =>
      IsolateTest(noop, {
        fieldName,
        testFn: jest.fn(),
        message,
      })
    ).reduce((existing, { id }) => {
      expect(existing.has(id)).toBe(false);
      existing.add(id);
      return existing;
    }, new Set<string>());
  });

  describe('testObject.warn', () => {
    it('Should mark the test as warning', () => {
      expect(VestTestInspector.warns(testObject)).toBe(false);
      VestTestMutator.warn(testObject);
      expect(VestTestInspector.warns(testObject)).toBe(true);
      expect(testObject).toMatchSnapshot();
    });
  });

  describe('testObject.fail', () => {
    beforeEach(() => {
      jest.resetModules();

      const { IsolateTest } = require('IsolateTest'); // eslint-disable-line @typescript-eslint/no-var-requires
      testObject = IsolateTest(noop, fieldName, jest.fn(), { message });
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('Should set status to failed', () => {
      expect(VestTestInspector.isFailing(testObject)).toBe(false);
      VestTestMutator.fail(testObject);
      expect(VestTestInspector.isFailing(testObject)).toBe(true);
    });
  });

  describe('testObject.valueOf', () => {
    test('When test did not fail', () => {
      expect(testObject.valueOf()).toBe(true);
    });

    test('When test failed', () => {
      VestTestMutator.fail(testObject);
      expect(testObject.valueOf()).toBe(false);
    });
  });

  describe('testObject.cancel', () => {
    it('Should set the testObject to cancel', () => {
      let testObject: IsolateTest;
      return TestPromise(done => {
        const suite = vest.create(() => {
          testObject = vest.test('f1', async () => {
            await wait(100);
          });
          vest.test('f2', async () => {
            await wait(100);
          });
          VestTestMutator.cancel(testObject);
        });
        suite();

        expect(VestTestInspector.isCanceled(testObject)).toBe(true);
        done();
      });
    });

    describe('final statuses', () => {
      let control = jest.fn();
      beforeEach(() => {
        control = jest.fn();
      });
      it('keep status unchanged when `failed`', () => {
        vest.create(() => {
          // async so it is not a final status
          const testObject = vest.test('f1', async () => {
            await wait(100);
          });
          VestTestMutator.fail(testObject);
          expect(VestTestInspector.isFailing(testObject)).toBe(true);
          VestTestMutator.skip(testObject);
          expect(VestTestInspector.isSkipped(testObject)).toBe(false);
          expect(VestTestInspector.isFailing(testObject)).toBe(true);
          VestTestMutator.cancel(testObject);
          expect(VestTestInspector.isCanceled(testObject)).toBe(false);
          expect(VestTestInspector.isFailing(testObject)).toBe(true);
          VestTestMutator.setPending(testObject);
          expect(VestTestInspector.isPending(testObject)).toBe(false);
          expect(VestTestInspector.isFailing(testObject)).toBe(true);
          control();
        })();
        expect(control).toHaveBeenCalledTimes(1);
      });

      it('keep status unchanged when `canceled`', () => {
        vest.create(() => {
          // async so it is not a final status
          const testObject = vest.test('f1', async () => {
            await wait(100);
          });
          VestTestMutator.cancel(testObject);
          expect(VestTestInspector.isCanceled(testObject)).toBe(true);
          VestTestMutator.fail(testObject);
          expect(VestTestInspector.isCanceled(testObject)).toBe(true);
          expect(VestTestInspector.isFailing(testObject)).toBe(false);
          VestTestMutator.skip(testObject);
          expect(VestTestInspector.isSkipped(testObject)).toBe(false);
          expect(VestTestInspector.isCanceled(testObject)).toBe(true);
          VestTestMutator.setPending(testObject);
          expect(VestTestInspector.isPending(testObject)).toBe(false);
          expect(VestTestInspector.isCanceled(testObject)).toBe(true);
          control();
        })();
        expect(control).toHaveBeenCalledTimes(1);
      });
    });
  });
});
