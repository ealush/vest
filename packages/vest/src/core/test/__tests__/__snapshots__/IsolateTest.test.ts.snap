// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsolateTest TestObject constructor 1`] = `
{
  "$type": "UnitTest",
  "children": [],
  "data": {
    "fieldName": "unicycle",
    "message": "I am Root.",
    "severity": "error",
    "status": "UNTESTED",
    "testFn": [MockFunction],
    "type": "Test",
  },
  "key": null,
  "keys": {},
  "output": null,
  "parent": null,
}
`;

exports[`IsolateTest testObject.warn Should mark the test as warning 1`] = `
{
  "$type": "UnitTest",
  "children": [],
  "data": {
    "fieldName": "unicycle",
    "message": "I am Root.",
    "severity": "warning",
    "status": "UNTESTED",
    "testFn": [MockFunction],
    "type": "Test",
  },
  "key": null,
  "keys": {},
  "output": null,
  "parent": null,
}
`;
